rules/scala/private/phases.bzl

```diff
@@ -0,0 +1,548 @@
+load(
+    "@rules_scala_annex//rules:providers.bzl",
+    _LabeledJars = "LabeledJars",
+    _ScalaConfiguration = "ScalaConfiguration",
+    _ScalaInfo = "ScalaInfo",
+    _ScalaRulePhase = "ScalaRulePhase",
+    _ZincConfiguration = "ZincConfiguration",
+    _ZincInfo = "ZincInfo",
+)
+load("//rules/common:private/utils.bzl", _collect = "collect", _write_launcher = "write_launcher")
+load(":private/import.bzl", _create_intellij_info = "create_intellij_info")
+
+def run_phases(ctx, phases):
+    scala_configuration = ctx.attr.scala[_ScalaConfiguration]
+    sdeps = java_common.merge(_collect(JavaInfo, scala_configuration.runtime_classpath + ctx.attr.deps))
+    init = struct(
+        scala_configuration = scala_configuration,
+        # todo: probably can remove this from init
+        sdeps = sdeps,
+    )
+
+    phase_providers = [p[_ScalaRulePhase] for p in ctx.attr.plugins if _ScalaRulePhase in p]
+    if phase_providers != []:
+        phases = phases[:]
+
+    for pp in phase_providers:
+        for (relation, peer_name, name, function) in pp.phases:
+            for idx, (needle, _) in enumerate(phases):
+                if needle == peer_name:
+                    if relation in ["-", "before"]:
+                        phases.insert(idx, (name, function))
+                    elif relation in ["+", "after"]:
+                        phases.insert(idx + 1, (name, function))
+
+    gd = {
+        "init": init,
+        "out": struct(
+            output_groups = {},
+            providers = [],
+        ),
+    }
+    g = struct(**gd)
+    for (name, function) in phases:
+        p = function(ctx, g)
+        if p != None:
+            gd[name] = p
+            g = struct(**gd)
+
+    return g
+
+_SINGLE_JAR_MNEMONIC = "SingleJar"
+
+#
+# PHASE: resources
+#
+# Resource files are merged into a zip archive.
+#
+# The output is returned in the jar field so the singlejar
+# phase will merge it into the final jar.
+#
+
+def phase_resources(ctx, g):
```

Comment:
Having a big comment block before each of these phases makes me think they should probably each be in their own file.

Reply:
I think so too. I have a few more changes in other PRs to iterate on the phase names and interfaces (in the global `g` struct) before I want to split them to separate files. I'd also like to organize the associated Scala source files along with the phases: the deps checker Scala can sit in the same area as the deps checker Starlark code.

Reply:
Looks good to me.